File: 0543-Diameter-of-Binary-Tree/0543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int: # type: ignore
        self.ans = 0

        def find_height(node):
            if not node: return 0

            left = find_height(node.left)
            right = find_height(node.right)
            height = max(left, right)

            # check the maxium diameter
            self.ans = max(self.ans, left + right)

            return 1 + height

        find_height(root)

        return self.ans

File: 0543-Diameter-of-Binary-Tree/0543.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    private int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        findHeight(root);
        return ans;
    }

    public int findHeight(TreeNode node) {
        if (node == null)
            return 0;

        int left = findHeight(node.left);
        int right = findHeight(node.right);
        int height = Math.max(left, right);

        ans = Math.max(ans, left + right);

        return 1 + height;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}